bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClient.java

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.xmppclient.internal;
+
+import org.jivesoftware.smack.*;
+import org.jivesoftware.smack.chat2.Chat;
+import org.jivesoftware.smack.chat2.ChatManager;
+import org.jivesoftware.smack.chat2.IncomingChatMessageListener;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.tcp.XMPPTCPConnection;
+import org.jivesoftware.smack.tcp.XMPPTCPConnectionConfiguration;
+
+import org.jxmpp.jid.EntityBareJid;
+import org.jxmpp.jid.impl.JidCreate;
+import org.jxmpp.stringprep.XmppStringprepException;
+import org.openhab.binding.xmppclient.handler.XMPPClientMessageSubscriber;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link XMPPClient} is lib foe handle XMPP connection and messagingю
+ *
+ * @author Pavel Gololobov - Initial contribution
+ */
+public class XMPPClient implements IncomingChatMessageListener, ConnectionListener {
+
+    private final Logger logger = LoggerFactory.getLogger(XMPPClient.class);
+
+    private AbstractXMPPConnection connection;
+
+    private ChatManager chatManager;
+
+    private Set<XMPPClientMessageSubscriber> subscribers = new HashSet<>();
+
+    public void subscribe(XMPPClientMessageSubscriber channel) {
+        logger.debug("Channel {} subscribed", channel.getName());
+        subscribers.add(channel);
+    }
+
+    public void unsubscribe(XMPPClientMessageSubscriber channel) {
+        logger.debug("Channel {} unsubscribed", channel.getName());
+        subscribers.remove(channel);
+    }
+
+    public void connect(String host, Integer port, String login, String password, String resource) {
+        disconnect();
+        try {
+            XMPPTCPConnectionConfiguration config = XMPPTCPConnectionConfiguration.builder()
+                    .setUsernameAndPassword(login, password)
```

Comment:
is this all null-safe? some of these fields are allowed to be null

Reply:
fixed